File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-bond-option-5-4/MakeWholeAmount.cpp
// MakeWholeAmount.cpp 
#include "MakeWholeAmount.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

MakeWholeAmount::MakeWholeAmount(TiXmlNode* xmlNode)
: SwapCurveValuation(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //interpolationMethodNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* interpolationMethodNode = xmlNode->FirstChildElement("interpolationMethod");

   if(interpolationMethodNode){interpolationMethodIsNull_ = false;}
   else{interpolationMethodIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- interpolationMethodNode , address : " << interpolationMethodNode << std::endl;
   #endif
   if(interpolationMethodNode)
   {
      if(interpolationMethodNode->Attribute("href") || interpolationMethodNode->Attribute("id"))
      {
          if(interpolationMethodNode->Attribute("id"))
          {
             interpolationMethodIDRef_ = interpolationMethodNode->Attribute("id");
             interpolationMethod_ = boost::shared_ptr<InterpolationMethod>(new InterpolationMethod(interpolationMethodNode));
             interpolationMethod_->setID(interpolationMethodIDRef_);
             IDManager::instance().setID(interpolationMethodIDRef_,interpolationMethod_);
          }
          else if(interpolationMethodNode->Attribute("href")) { interpolationMethodIDRef_ = interpolationMethodNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { interpolationMethod_ = boost::shared_ptr<InterpolationMethod>(new InterpolationMethod(interpolationMethodNode));}
   }

   //earlyCallDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* earlyCallDateNode = xmlNode->FirstChildElement("earlyCallDate");

   if(earlyCallDateNode){earlyCallDateIsNull_ = false;}
   else{earlyCallDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- earlyCallDateNode , address : " << earlyCallDateNode << std::endl;
   #endif
   if(earlyCallDateNode)
   {
      if(earlyCallDateNode->Attribute("href") || earlyCallDateNode->Attribute("id"))
      {
          if(earlyCallDateNode->Attribute("id"))
          {
             earlyCallDateIDRef_ = earlyCallDateNode->Attribute("id");
             earlyCallDate_ = boost::shared_ptr<IdentifiedDate>(new IdentifiedDate(earlyCallDateNode));
             earlyCallDate_->setID(earlyCallDateIDRef_);
             IDManager::instance().setID(earlyCallDateIDRef_,earlyCallDate_);
          }
          else if(earlyCallDateNode->Attribute("href")) { earlyCallDateIDRef_ = earlyCallDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { earlyCallDate_ = boost::shared_ptr<IdentifiedDate>(new IdentifiedDate(earlyCallDateNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<InterpolationMethod> MakeWholeAmount::getInterpolationMethod()
{
   if(!this->interpolationMethodIsNull_){
        if(interpolationMethodIDRef_ != ""){
             return boost::shared_static_cast<InterpolationMethod>(IDManager::instance().getID(interpolationMethodIDRef_));
        }else{
             return this->interpolationMethod_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<InterpolationMethod>();
   }
}
boost::shared_ptr<IdentifiedDate> MakeWholeAmount::getEarlyCallDate()
{
   if(!this->earlyCallDateIsNull_){
        if(earlyCallDateIDRef_ != ""){
             return boost::shared_static_cast<IdentifiedDate>(IDManager::instance().getID(earlyCallDateIDRef_));
        }else{
             return this->earlyCallDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<IdentifiedDate>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-bond-option-5-4/MakeWholeAmount.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class MakeWholeAmount : SwapCurveValuation
    {
        public MakeWholeAmount(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNode interpolationMethodNode = xmlNode.SelectSingleNode("interpolationMethod");
            
            if (interpolationMethodNode != null)
            {
                if (interpolationMethodNode.Attributes["href"] != null || interpolationMethodNode.Attributes["id"] != null) 
                {
                    if (interpolationMethodNode.Attributes["id"] != null) 
                    {
                        interpolationMethodIDRef_ = interpolationMethodNode.Attributes["id"].Value;
                        InterpolationMethod ob = new InterpolationMethod(interpolationMethodNode);
                        IDManager.SetID(interpolationMethodIDRef_, ob);
                    }
                    else if (interpolationMethodNode.Attributes["href"] != null)
                    {
                        interpolationMethodIDRef_ = interpolationMethodNode.Attributes["href"].Value;
                    }
                    else
                    {
                        interpolationMethod_ = new InterpolationMethod(interpolationMethodNode);
                    }
                }
                else
                {
                    interpolationMethod_ = new InterpolationMethod(interpolationMethodNode);
                }
            }
            
        
            XmlNode earlyCallDateNode = xmlNode.SelectSingleNode("earlyCallDate");
            
            if (earlyCallDateNode != null)
            {
                if (earlyCallDateNode.Attributes["href"] != null || earlyCallDateNode.Attributes["id"] != null) 
                {
                    if (earlyCallDateNode.Attributes["id"] != null) 
                    {
                        earlyCallDateIDRef_ = earlyCallDateNode.Attributes["id"].Value;
                        IdentifiedDate ob = new IdentifiedDate(earlyCallDateNode);
                        IDManager.SetID(earlyCallDateIDRef_, ob);
                    }
                    else if (earlyCallDateNode.Attributes["href"] != null)
                    {
                        earlyCallDateIDRef_ = earlyCallDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        earlyCallDate_ = new IdentifiedDate(earlyCallDateNode);
                    }
                }
                else
                {
                    earlyCallDate_ = new IdentifiedDate(earlyCallDateNode);
                }
            }
            
        
        }
        
    
        #region InterpolationMethod_
        private InterpolationMethod interpolationMethod_;
        public InterpolationMethod InterpolationMethod_
        {
            get
            {
                if (this.interpolationMethod_ != null)
                {
                    return this.interpolationMethod_; 
                }
                else if (this.interpolationMethodIDRef_ != null)
                {
                    interpolationMethod_ = IDManager.getID(interpolationMethodIDRef_) as InterpolationMethod;
                    return this.interpolationMethod_; 
                }
                else
                {
                      return this.interpolationMethod_; 
                }
            }
            set
            {
                if (this.interpolationMethod_ != value)
                {
                    this.interpolationMethod_ = value;
                }
            }
        }
        #endregion
        
        public string interpolationMethodIDRef_ { get; set; }
        #region EarlyCallDate_
        private IdentifiedDate earlyCallDate_;
        public IdentifiedDate EarlyCallDate_
        {
            get
            {
                if (this.earlyCallDate_ != null)
                {
                    return this.earlyCallDate_; 
                }
                else if (this.earlyCallDateIDRef_ != null)
                {
                    earlyCallDate_ = IDManager.getID(earlyCallDateIDRef_) as IdentifiedDate;
                    return this.earlyCallDate_; 
                }
                else
                {
                      return this.earlyCallDate_; 
                }
            }
            set
            {
                if (this.earlyCallDate_ != value)
                {
                    this.earlyCallDate_ = value;
                }
            }
        }
        #endregion
        
        public string earlyCallDateIDRef_ { get; set; }
        
    
        
    
    }
    
}

